### PR TITLE
Rewrite sync and network layers to use futures

### LINKED ISSUE #834: Rewrite sync and network layers to use futures
Right now the network layer spawns a thread that runs additional tokio threads. That is because the sync layer still uses the old devp2p API based on callbacks. The code should be reworked to not use them.

### PR DESCRIPTION
**Pull request by tomaka**
Fix #834 
cc #2536 

Removes the background thread of the network crate.
Creating a network service now returns a `(Arc<Network>, NetworkMut)`.

The `NetworkMut` implements `Future` and must be driven to completion in order for the network to work. It corresponds to what the background thread previously had on its stacks.

Also includes a second commit for #2416 that makes us depend on `tokio-timer` instead of `tokio` for timers.

### CODE REVIEW

File: core/network/src/service.rs [hunk 1]
```diff
@@ -521,136 +519,83 @@ impl<B: BlockT, F: FnOnce(&mut ConsensusGossip<B>, &mut Context<B>)> GossipTask<
...
+	type Item = ();
+	type Error = io::Error;
 
-	// Interval at which we send status updates on the `status_sinks`.
-	let mut status_interval = tokio::timer::Interval::new_interval(STATUS_INTERVAL);
-	// Interval at which we update the `connected_peers` Arc.
-	let mut connected_peers_interval = tokio::timer::Interval::new_interval(CONNECTED_PEERS_INTERVAL);
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		// Implementation of `protocol::NetworkOut` using the available local variables.
+		struct Ctxt<'a, B: BlockT>(&'a mut NetworkService<Message<B>>, &'a PeersetHandle);
```
Comment (arkpar) [inline core/network/src/service.rs]: This is an old code, but now is a good time to rename it to something readable according to our guidelines, like "Context"

**Comment by arkpar** [comment]
`Service` and `NetworkMut` are a bit confusing. Looks lie `Service` is effectively just an interface to the networking. I'd rename `Service` -> `ServiceHandle` and `NetworkMut` -> `Service`. `Service::new` would only return `Service` object that implements `Future` and also has a `handle` function that returns `ServiceHandle`.  Do you think this makes sense?

**Comment by tomaka** [comment]
I agree with the naming issue, but changing `Network` to something else would break almost all the other crates of Substrate.
If that's fine in terms of backwards compatibility, I'll do it.

**Comment by arkpar** [comment]
We don't really maintain rust API stability yet. This PR already breaks the API anyway.

**Comment by tomaka** [comment]
I went with `NetworkService` and `NetworkWorker`, as `NetworkSeviceWorker` looked a bit long, if that's ok.

**Comment by arkpar** [comment]
How about my suggestion to rename `Ctxt` as well and split `NetworkService::new` into `NetworkWorker::new` and `NetworkWorker::service`?

**Comment by tomaka** [comment]
Done. I also moved `new` to `NetworkWorker`. Since it returns a `NetworkWorker` it makes more sense to do so.

**Pull request merged**